community/en/ImageSegmentation_ModelSubclassing.ipynb

```diff
@@ -0,0 +1,732 @@
+        "    # define layers\n",
+        "    self.layer_1 = layers.Activation('relu')\n",
+        "    self.layer_2 = layers.SeparableConv2D(self.filter_size, 3, padding='same')\n",
+        "    self.layer_3 = layers.BatchNormalization()\n",
+        "\n",
+        "    self.layer_4 = layers.Activation('relu')\n",
+        "    self.layer_5 = layers.SeparableConv2D(self.filter_size, 3, padding='same')\n",
+        "    self.layer_6 = layers.BatchNormalization()\n",
+        "\n",
+        "    self.layer_7 = layers.MaxPooling2D(3, strides=2, padding='same')\n",
+        "\n",
+        "    # project residual\n",
+        "    self.residual_layer = layers.Conv2D(self.filter_size, 1, strides=2, padding='same')\n",
+        "  def call(self, inputs):\n",
+        "    x = self.layer_1(inputs)\n",
+        "    x = self.layer_2(x)\n",
+        "    x = self.layer_3(x)\n",
+        "    x = self.layer_4(x)\n",
+        "    x = self.layer_5(x)\n",
+        "    x = self.layer_6(x)\n",
+        "    x = self.layer_7(x)\n",
+        "    residual = self.residual_layer(inputs)\n",
+        "    x = layers.add([x, residual])\n",
+        "    return x\n",
+        "\n",
+        "class DecoderBlock(tf.keras.Model):\n",
+        "  def __init__(self, filter_size):\n",
+        "    # initilize instance variables\n",
+        "    super(DecoderBlock, self).__init__()\n",
+        "    self.filter_size = filter_size\n",
+        "    # define layers\n",
+        "    self.layer_1 = layers.Activation('relu')\n",
+        "    self.layer_2 = layers.Conv2DTranspose(self.filter_size, 3, padding='same')\n",
+        "    self.layer_3 = layers.BatchNormalization()\n",
+        "\n",
+        "    self.layer_4 = layers.Activation('relu')\n",
+        "    self.layer_5 = layers.Conv2DTranspose(self.filter_size, 3, padding='same')\n",
+        "    self.layer_6 = layers.BatchNormalization()\n",
+        "\n",
+        "    self.layer_7 = layers.UpSampling2D(2)\n",
+        "\n",
+        "    # project residual\n",
+        "    self.residual_layer_1 = layers.UpSampling2D(2)\n",
+        "    self.residual_layer_2 = layers.Conv2D(filter_size, 1, padding='same')\n",
+        "  def call(self, inputs):\n",
+        "    x = self.layer_1(inputs)\n",
+        "    x = self.layer_2(x)\n",
+        "    x = self.layer_3(x)\n",
+        "    x = self.layer_4(x)\n",
+        "    x = self.layer_5(x)\n",
+        "    x = self.layer_6(x)\n",
+        "    x = self.layer_7(x)\n",
+        "    residual = self.residual_layer_1(inputs)\n",
+        "    residual = self.residual_layer_2(residual)\n",
+        "\n",
+        "    x = layers.add([x, residual])\n",
+        "    return x"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "EGP8hTwhc3Ww",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "class UnetModel(tf.keras.Model):\n",
```

Comment:
Consider changing the name since Unet refers to this paper: https://arxiv.org/pdf/1505.04597.pdf

So the name "Unet" is, AFAIKT, about those cross-over connections.

Reply:
I see. Changed the name to ImageSegmentationModel. Thought UNet means the general encoder-decoder type architecture, did not realize the cross-over connections. Thanks for the insight.